This folder contains all scripts related to the richter transformation dataset. Here there is a brief description of the scripts.

model_label_transfer.R: This script provides a comprehensive workflow for analyzing scRNA-seq data using the matchSCore2 package in R. It starts by identifying marker genes, scales the data using these markers, trains a classification model, and projects new cells onto the trained model to predict their identities. 

richter_transformation.R: This script provides a comprehensive analysis of the single-cell RNA sequencing data from patients analyzed in the Nadeu, F., Royo, R., Massoni-Badosa, R. et al. paper about Richter Transformation, focusing on the ZAP70 gene expression. It includes quality control checks and differential expression analysis to identify genes associated with ZAP70 expression status (ZAP+ vs ZAP-). Gene ontology (GO) enrichment analysis was performed on differentially expressed genes. The script also integrates and analyzes multiple samples, annotating cells and visualizing clustering and ZAP70 expression across samples using dimensionality reduction techniques like UMAP. Finally, it identifies and saves top marker genes for each cluster associated with ZAP70 expression status across different samples.
